torch/ao/quantization/quantize.py

```diff
@@ -261,7 +277,7 @@ def prepare(model, inplace=False, allow_list=None,
     """
     torch._C._log_api_usage_once("quantization_api.quantize.prepare")
     if prepare_custom_config_dict is None:
-        prepare_custom_config_dict = {}
+        prepare_custom_config_dict = get_default_custom_config_dict()
```

Comment:
I think it's better to keep the defaults as well. Something like this:

```python
_prepare_custom_config_dict = get_default_custom_config_dict()
_prepare_custom_config_dict.update(prepare_custom_config_dict or {})
prepare_custom_config_dict = _prepare_custom_config_dict

for key, value in prepare_custom_config_dict['float_to_observed_custom_module_class').items():
    if value is None:
         prepare_custom_config_dict['float_to_observed_custom_module_class').pop(key)

for key, value in prepare_custom_config_dict['observed_to_quantized_custom_module_class').items():
    if value is None:
         prepare_custom_config_dict['observed_to_quantized_custom_module_class').pop(key)
```

(Of course you can rewrite it to avoid the temporary `_prepare_custom_config_dict`).


The benefit of it is that if the user specifies custom config dict that does not involve the LSTM or MHA, we still want to support them:

```python
custom_module_config = {
    'float_to_observed_custom_module_class': {
        MyModule: MyOtherModule
    },
    'observed_to_quantized_custom_module_class': {
        MyOtherModule: MyOtherOtherModule
    }
}
prepared = torch.ao.quantization.prepare(model, prepare_custom_config_dict=custom_module_config)
```

In the above example, if the `model` has an instance of an `nn.LSTM`, we still want to be able to replace it.

The downside of using it by default is that if the user doesn't want to quantize the LSTM, they would need to be explicit about it:

```python
custom_module_config = {
    'float_to_observed_custom_module_class': {
        torch.nn.LSTM: None
    },
    'observed_to_quantized_custom_module_class': {
        torch.nn.quantizable.LSTM: None
    }
}
prepared = torch.ao.quantization.prepare(model, prepare_custom_config_dict=custom_module_config)
```

What do you think?

Reply:
Per the proposal (https://docs.google.com/document/d/16upncHr8zLQaynosbskCZP522oTkEMWzioMHKp1x0_s/edit), one of the goals was to align with the other quantization apis i.e. user specified matching overwrites rather than updates the defaults.

Reply:
sgtm